### ISSUE TITLE
Please build a universal binary for macOS

### ISSUE THREAD

**Issue by johnblommers** [jgm/pandoc#6960]
Are there any plans to produce a universal binary of Pandoc for the new Apple Silicon M1 Macintosh machines. These are ARM64 based computers.

While Pandoc rendered in Intel might work under the Rosetta 2 translator I believe users will appreciate the performance gains from an ARM64 binary.

**Comment by jgm** [comment]
It's not something we can do right now (the infrastructure isn't there -- I don't think there's a version of ghc that targets that architecture -- and even then we'd have to have a build machine that could run it on travis).  But we can keep this open for the future.

Perhaps someone can report here on their experiences running pandoc on an M1 machine in emulation mode.

**Comment by ickc** [comment]
It is an "upstream" issue, relevant links:

* [Does Haskell work on Apple silicon?](https://isapplesiliconready.com/app/Haskell)
* [Support for Apple ARM hardware (#18664) · Issues · Glasgow Haskell Compiler / GHC · GitLab](https://gitlab.haskell.org/ghc/ghc/-/issues/18664)

where the latest report from there:

> I've managed to compile a GHC following the above instructions.  I've then ran the test suite.  Most of the failures output:

Suggestions:

- change title to "macOS Apple Silicon package" or "osx-arm64 package"
- apply label: "platform:osx", "package", "status:waiting-for-upstream"

**Comment by nonoesp** [comment]
👋🏻 Hi, everyone!

I found this issue as I was setting up my MacBook Pro (13-inch, M1, 2020) and thought I'd give this a try.

I hope this is helpful. =)

TL;DR: It works with Rosetta.

## Install Homebrew with Rosetta

You can create an installation of Homebrew specific to run with Rosetta bottles.

```bash
# Install Homebrew for x86_64 architecture
# https://soffes.blog/homebrew-on-apple-silicon
arch -x86_64 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
```

## Install pandoc

```
# Install pandoc using that version of Homebrew
arch -x86_64 brew install pandoc
```

Outputs

```
==> Downloading https://homebrew.bintray.com/bottles/pandoc-2.11.4.big_sur.bottle.tar.gz
Already downloaded: /Users/nono/Library/Caches/Homebrew/downloads/34e1528919e624583d70b1ef24381db17f730fc69e59144bf48abedc63656678--pandoc-2.11.4.big_sur.bottle.tar.gz
==> Pouring pandoc-2.11.4.big_sur.bottle.tar.gz
🍺  /usr/local/Cellar/pandoc/2.11.4: 10 files, 146.0MB
```

```
# Check pandoc's version
arch -x86_64 pandoc --version
```

Outputs

```
pandoc 2.11.4
Compiled with pandoc-types 1.22, texmath 0.12.1, skylighting 0.10.2,
citeproc 0.3.0.5, ipynb 0.1.0.1
User data directory: /Users/nono/.local/share/pandoc or /Users/nono/.pandoc
Copyright (C) 2006-2021 John MacFarlane. Web:  https://pandoc.org
This is free software; see the source for copying conditions. There is no
warranty, not even for merchantability or fitness for a particular purpose.
```

## Convert Markdown to Html

```
arch -x86_64 pandoc sample.md -o sample.html
```

Contents of sample.md:

```
# Hello, Apple Silicon!

- Pandoc
- seems
- to
- work.
```

Contents of sample.html:

```
<h1 id="hello-apple-silicon">Hello, Apple Silicon!</h1>
<ul>
<li>Pandoc</li>
<li>seems</li>
<li>to</li>
<li>work.</li>
</ul>
```

Cheers,
Nono

**Comment by shedd** [comment]
Thank you for this! 

Confirmed this does work - the only issue that I ran into was needing to specify the full path to `brew` in the installation command: `arch -x86_64 /usr/local/bin/brew install pandoc`

**Comment by Linus00012** [comment]
> Thank you for this!
> 
> Confirmed this does work - the only issue that I ran into was needing to specify the full path to `brew` in the installation command: `arch -x86_64 /usr/local/bin/brew install pandoc`

Check out for this [article](https://soffes.blog/homebrew-on-apple-silicon)!

**Comment by kerim** [comment]
Any developments on this? I really miss the ease of use of the PanDoc app on my mac...

**Comment by kkbt** [comment]
As a preliminary solution, I was successful in installing the Intel version of homebrew in /usr/local (in parallel to the M1 version in /opt/homebrew) using Apple's Rosetta 2 layer. The Intel packages (homebrew formulae) seem to work without any problem on the Apple M1 architecture. Both pdftk and pandoc work even without prefixing 'arch -x86_64' (e.g., the command 'pandoc sample.md -o sample.html' as in the example linked below).


Commands:

arch -x86_64 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
alias ibrew='arch -x86_64 /usr/local/bin/brew'
ibrew analytics off
ibrew install pdftk-java
ibrew install pandoc


Further information:

- [Homebrew on Apple Silicon](https://soffes.blog/homebrew-on-apple-silicon)
- [Install Homebrew with Rosetta](https://github.com/jgm/pandoc/issues/6960#issuecomment-778698895)
- [Main thread](https://github.com/Homebrew/formulae.brew.sh/issues/467)

**Comment by yorkofyou** [comment]
Haskell Compiler (GHC) 8.10.5 released today with native Apple Silicon support.

**Comment by jgm** [comment]
Now we have the compiler, we need to figure out how to build the package.
Currently we build the mac packages on CircleCI -- so unless they are offering M1 build machines (or unless ghc can cross-compile), we're still not there...

**Comment by jgm** [comment]
Circle CI does not yet support M1 build machines.
https://support.circleci.com/hc/en-us/articles/360056461452-Apple-M1-Apple-Silion-Support-on-CircleCI

However, we may be able to set up a build on ec2 (as we currently do for linux arm), once M1 instances are available on EC2 (projected for some time in 2021).

**Comment by jgm** [comment]
Instructions for using lipo to create a universal binary:
https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary

**Comment by carlocab** [comment]
ICYMI M1-native binaries of `pandoc` have been available from Homebrew for about a week now. CC @storopoli

**Comment by storopoli** [comment]
Thanks! Amazing `brew install pandoc` works!

**Comment by nonoesp** [comment]
Thanks! This is great news! 🥳

**Comment by cbrnr** [comment]
FWIW, Cirrus CI offers ARM64 Macs which could be used to build a native package for Apple Silicon. We've been using it in https://github.com/mne-tools/mne-python in case you are interested.

**Comment by tarleb** [comment]
Thanks @cbrnr, that is valuable info!

**Comment by jgm** [comment]
Here's their page on the m1 mac vms:
https://cirrus-ci.org/guide/macOS/

**Comment by jgm** [comment]
Some success here: https://cirrus-ci.com/task/6747943965818880

**Comment by cbrnr** [comment]
Some success? It's working, right? 🚀

**Comment by jgm** [comment]
I've got it working pretty well now, I think.  Henceforth we'll build macos arm, linux arm, and linux amd releases on Cirrus, and macos intel and windows on GitHub.  My thought is not to provide a universal binary (I'm actually not sure how to do that) but just to provide separate packages for intel and apple silicon...

**Comment by carlocab** [comment]
You can just combine your Intel and Apple Silicon builds using something like
```sh
lipo -create -output pandoc /path/to/pandoc-x86_64 /path/to/pandoc-arm64
```
This will produce a universal `pandoc` in your current working directory.

**Comment by cbrnr** [comment]
I think separate builds are fine (Apple Silicon machines don't need Intel code and vice versa).

May I ask what the first official release with native Apple Silicon support will be?

**Comment by jgm** [comment]
Next release will have it.  I don't know when that will be exactly.

**Comment by cbrnr** [comment]
Thank you @jgm!
